fastn-core/src/migrations.rs

```diff
@@ -0,0 +1,224 @@
+                .to_json_string(&doc, false)?
+                .unwrap();
+            let migration_name = fastn_core::utils::generate_hash(&query);
+            let migration_number = (idx + 1) as i64;
+            migration_data.push(MigrationDataFTD {
+                number: migration_number,
+                name: migration_name,
+                content: query,
+            });
+        } else {
+            unimplemented!(); // todo: throw error
+        }
+    }
+
+    migration_data
+}
+
+fn is_fastn_migration(component: &ftd::interpreter::Component) -> bool {
+    component.name.eq("fastn#migration") || component.name.eq("fastn.migration")
+}
+
+struct MigrationDataSQL {
+    number: i64,
+    name: String,
+}
+
+async fn find_migrations_to_apply(
+    migration_ftd: &[MigrationDataFTD],
+    migration_sql: &[MigrationDataSQL],
+) -> Vec<MigrationDataFTD> {
+    if migration_sql.len() > migration_ftd.len() {
+        unreachable!("Can't delete the applied migration") // todo: throw error
+    }
+
+    let migration_sql: std::collections::HashMap<i64, MigrationDataSQL> = migration_sql
+        .into_iter()
+        .map(|val| (*val.number, val.clone()))
+        .collect();
+    let mut migrations_to_apply = vec![];
+    for m_ftd in migration_ftd {
+        match migration_sql.get(*m_ftd.number) {
+            Some(m_sql) => {
+                if m_sql.name.ne(&m_ftd.name) {
+                    unreachable!("Cannot change the content of migration") // todo: throw error
+                }
+            }
+            None => {
+                migrations_to_apply.push(m_ftd.clone());
+            }
+        }
+    }
+    migrations_to_apply
+}
+
+async fn get_migration_from_sql(
+    config: &fastn_core::Config,
+    db: &str,
+) -> Result<Vec<MigrationDataSQL>, fastn_utils::SqlError> {
+    let results = config
+        .ds
+        .sql_query(
+            db,
+            format!(
+                r#"
+                    SELECT
+                        migration_number, migration_name
+                    FROM
+                        fastn_migration
+                    WHERE
+                        app_name='{}'
+                "#,
+                config.package.name
+            ),
+            vec![],
+        )
+        .await?;
+
+    let mut migration_details = vec![];
+
+    for entry in results {
+        let migration_number: i64 = serde_json::from_value(
+            entry
+                .get(0)
+                .expect("fastn_migration::migration_number not found.")
+                .clone(),
+        )
+        .unwrap();
+        let migration_name: String = serde_json::from_value(
+            entry
+                .get(1)
+                .expect("fastn_migration::migration_name not found.")
+                .clone(),
+        )
+        .unwrap();
+        migration_details.push(MigrationDataSQL {
+            number: migration_number,
+            name: migration_name,
+        });
+    }
```

Comment:
This is to check if migraiton names etc have not changed. Let’s not do this, as this is too much work on every page load. Find the latest migration number, and compare it with number of migrations in fastn.ftd file, and if they differ, and it comes time to apply an actual migration, we first do rigerous check and then apply migraiton.